utils/backoff.go

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) 2016-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package utils
+
+import (
+	"time"
+)
+
+const (
+	backoffBase uint64 = 128
+	maxAttempts        = 4
+)
+
+// BackoffOperation is executed by Retry.
+// The BackoffOperation will be retried the provided
+// number attempts if returning an error.
+type BackoffOperation func() error
+
+// ProgressiveRetry executes a BackoffOperation and retries the operation 3 times upon error.
+func ProgressiveRetry(bo BackoffOperation) error {
+	var t *time.Timer
+	var attempt uint64
+
+	for {
+		err := bo()
+		if err == nil {
+			return nil
+		}
+
+		nextRetry := NextRetry(attempt)
+		if t == nil {
+			t = time.NewTimer(nextRetry)
+		} else {
+			t.Reset(nextRetry)
+		}
+
+		attempt++
+		if attempt >= maxAttempts {
+			return err
+		}
+
+		// Wait until timer is finished before trying again
+		<-t.C
```

Comment:
What's the advantage to using a timer here instead of just `time.Sleep`?

Reply:
@jesperhansen17, wanted to ping again re: `time.Sleep`. From what I can see, there's no reason to create a timer for this case. It would make sense if we wanted to wait on multiple channels to support, say, an abort, but a `time.Sleep` seems like it would simplify the code nicely.

Reply:
In this implementation I can't say there is any, but I used timer before and that's why I choose to use it again. What I do like with it comes as a channel and therefore I can make use of other go routines and I feel it gives better flexibility in handling parallel go routines.

But yeah, here in this case I can't really say what the main advantage is.